zebra-chain/src/notes/sprout.rs

```diff
@@ -1,33 +1,88 @@
 //!
+#![allow(clippy::unit_arg)]
 #![allow(dead_code)]
 
-use serde::{Deserialize, Serialize};
-use std::{
-    fmt,
-    io::{self},
-};
+use std::{fmt, io};
 
-#[cfg(test)]
-use proptest::{collection::vec, prelude::*};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
 
-use crate::serde_helpers;
-use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
+use crate::{
+    keys::sprout::PayingKey,
+    serde_helpers,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    types::amount::{Amount, NonNegative},
+};
 
 use super::{memo::Memo, *};
 
+/// Nullifier seed, named rho in the [spec][ps].
 ///
+/// [ps]: https://zips.z.cash/protocol/protocol.pdf#sproutkeycomponents
+#[derive(Clone, Copy)]
+struct NullifierSeed([u8; 32]);
+
+impl AsRef<[u8]> for NullifierSeed {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+/// A Nullifier Set for Sprout transactions
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct Nullifier([u8; 32]);
+
+impl From<[u8; 32]> for Nullifier {
+    fn from(buf: [u8; 32]) -> Self {
+        Self(buf)
+    }
+}
+
+impl ZcashDeserialize for Nullifier {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        let bytes = reader.read_32_bytes()?;
+
+        Ok(Self(bytes))
+    }
+}
+
+impl ZcashSerialize for Nullifier {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&self.0[..])
+    }
+}
+
+/// A Note represents that a value is spendable by the recipient who
+/// holds the spending key corresponding to a given shielded payment
+/// address.
 pub struct Note {
-    // TODO: refine type as a SHA-256d output derived from a spending key.
-    paying_key: [u8; 32],
-    value: u64,
-    // TODO: refine type as the input to the PRF that results in a nullifier.
-    nullifier_seed: [u8; 32],
+    paying_key: PayingKey,
+    value: Amount<NonNegative>,
+    nullifier_seed: NullifierSeed,
     note_commitment_randomness: NoteCommitmentRandomness,
 }
 
+impl Note {
+    pub fn commitment(&self) -> NoteCommitment {
```

Comment:
@hdevalence thoughts on naming here? Should this method use an active verb ('`commit`') when it doesn't change the state of the struct? In comparison, librustzcash has `cm()` as a method on their `Note`.

Reply:
I think an active verb is fine because the method is computing the commitment, and the &self parameter is a big clue that there's no change to the state of the struct.  But commitment is also fine.  I think either are preferable to `cm`.